src/reference/asciidoc/kafka.adoc

```diff
@@ -1271,6 +1271,31 @@ public void listen(List<Foo> foos, @Header(KafkaHeaders.OFFSET) List<Long> offse
 
 Notice that you can still access the batch headers too.
 
+Starting with _versions 1.3.3_ the `org.springframework.core.convert.ConversionService` used by the default
```

Comment:
I don't think that we are going to back port it down into the `1.3.x`. (`2.1.1` should be good)
There is a workaround like `MessageHandlerMethodFactory` injection into the `KafkaListenerEndpointRegistrar` as you correctly mention below.

There should be `,` after version.

Reply:
I'm aware of the workaround, as i'm currently forced to use it. :-)

But I think the beauty of my solution is that unless you have to programatically configure any other properties of the listener handling that aren't confiurable through the `KafkaProperties`, you basically just have to put `@EnableKafka` somewhere and implement `Converter` for your incoming message types. Or maybe I'm misunderstanding the workaround, because this is what I have to do, to achieve the same behaviour:

``` java
@Component
public class KafkaConfig implements KafkaListenerConfigurer {

    @Autowired
    BeanFactory beanFactory;

    @Autowired
    ConversionService conversionService;

    @Override
    public void configureKafkaListeners(KafkaListenerEndpointRegistrar registrar) {
        registrar.setMessageHandlerMethodFactory(createDefaultMessageHandlerMethodFactory());
    }

    private MessageHandlerMethodFactory createDefaultMessageHandlerMethodFactory() {
        DefaultMessageHandlerMethodFactory defaultFactory = new DefaultMessageHandlerMethodFactory();
        defaultFactory.setBeanFactory(beanFactory);

        ConfigurableBeanFactory cbf =
                (beanFactory instanceof ConfigurableBeanFactory ?
                        (ConfigurableBeanFactory) beanFactory : null);

        defaultFactory.setConversionService(conversionService);

        List<HandlerMethodArgumentResolver> argumentResolvers = new ArrayList<>();

        argumentResolvers.add(new HeaderMethodArgumentResolver(conversionService, cbf));
        argumentResolvers.add(new HeadersMethodArgumentResolver());

        final GenericMessageConverter messageConverter = new GenericMessageConverter(conversionService);
        argumentResolvers.add(new MessageMethodArgumentResolver(messageConverter));
        argumentResolvers.add(new PayloadArgumentResolver(messageConverter) {

            @Override
            protected boolean isEmptyPayload(Object payload) {
                return payload == null || payload instanceof KafkaNull;
            }

        });
        defaultFactory.setArgumentResolvers(argumentResolvers);

        defaultFactory.afterPropertiesSet();
        return defaultFactory;
    }

}
```

Reply:
Seems for me your workaround is overhead.
That should be like this:
```
@Override
public void configureKafkaListeners(KafkaListenerEndpointRegistrar registrar) {
        registrar.setMessageHandlerMethodFactory(defaultMessageHandlerMethodFactory());
}

@Bean
public MessageHandlerMethodFactory defaultMessageHandlerMethodFactory() {
    DefaultMessageHandlerMethodFactory defaultFactory = new DefaultMessageHandlerMethodFactory();
   defaultFactory.setConversionService(conversionService);
   return defaultFactory;
```

Reply:
Yes, I noticed that too. But this way it lacks the handling of `KafkaNull`. Which shouldn't cause any problems in my case. I'm also currently looking into adding a `Validator` to the `MessageHandlerMethodFactory'. Something you should consider adding in similar fashion some day. :-D

Reply:
Oh! I see what you mean about `KafkaNull`. Hard to read long code on GH 😄 .

Well, for that purpose we have `DefaultMessageHandlerMethodFactory.setCustomArgumentResolvers()`.